library/objective-c/EnvoyEngineImpl.m

```diff
@@ -193,19 +195,18 @@ - (void)recordCounter:(NSString *)elements count:(NSUInteger)count {
 
 - (void)startObservingLifecycleNotifications {
   NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
-  [notificationCenter addObserver:self
-                         selector:@selector(lifecycleDidChangeWithNotification:)
-                             name:UIApplicationWillResignActiveNotification
-                           object:nil];
   [notificationCenter addObserver:self
                          selector:@selector(lifecycleDidChangeWithNotification:)
                              name:UIApplicationWillTerminateNotification
                            object:nil];
 }
 
 - (void)lifecycleDidChangeWithNotification:(NSNotification *)notification {
-  NSLog(@"[Envoy] triggering stats flush (%@)", notification.name);
-  flush_stats();
+  NSLog(@"[Envoy] terminating engine (%@)", notification.name);
+  // TODO: ensure this is called with the correct envoy_engine_t once multiple Engine objects can
+  // be allocated.
+  // https://github.com/lyft/envoy-mobile/issues/332
+  terminate_engine(0);
```

Comment:
Won't this terminate the engine on _any_ lifecycle notification? Or are we only registering for backgrounding/shutdown.

Reply:
Also, we should have access to the engine handle here. No need to hardcode `0`.

Reply:
I deleted the observer for `UIApplicationWillResignActiveNotification` so it will only be on termination. ack on the engine handle, will update.

Reply:
I think the notification actually has the lifecycle event on it, so we could be more picky here. My worry is that if someone adds it back in, it won't be obvious they're doing something wrong, and it could cause networking to fail until the app is restarted.